pallets/staking/src/lib.rs

```diff
@@ -2607,35 +2611,26 @@ decl_module! {
     }
 }
 
-impl<T: Trait> Module<T> {
+impl<T: Trait> OffchainInterface<T::AccountId> for Module<T> {
     /// POLYMESH-NOTE: This change is polymesh specific to query the list of all invalidate nominators
     /// It is recommended to not call this function on-chain. It is a non-deterministic function that is
     /// suitable for off-chain workers only.
-    pub fn fetch_invalid_cdd_nominators(buffer: u64) -> Vec<T::AccountId> {
-        let invalid_nominators = <Nominators<T>>::iter()
-            .filter_map(|(nominator_stash_key, _nominations)| {
-                if let Some(nominate_identity) =
-                    <identity::Module<T>>::get_identity(&(nominator_stash_key))
-                {
-                    if (<identity::Module<T>>::fetch_cdd(
-                        nominate_identity,
-                        buffer.saturated_into::<T::Moment>(),
-                    ))
+    fn fetch_invalid_cdd_nominators(buffer: u64) -> Vec<T::AccountId> {
+        <Nominators<T>>::iter()
+            .map(|(stash, _n)| (<Identity<T>>::get_identity(&stash), stash))
+            .filter(|(id, _s)| id.is_some())
+            .filter(|(id, _s)| {
+                (<Identity<T>>::fetch_cdd(id.unwrap(), buffer.saturated_into::<T::Moment>()))
                     .is_none()
-                    {
-                        return Some(nominator_stash_key);
-                    }
-                }
-                None
             })
-            .collect::<Vec<T::AccountId>>();
-        invalid_nominators
+            .map(|(_id, stash)| stash)
+            .collect::<Vec<T::AccountId>>()
     }
 
     /// POLYMESH-NOTE: This is Polymesh specific change.
     /// Here we are assuming that passed targets are always be a those nominators whose cdd
     /// claim get expired or going to expire after the `buffer_time`.
-    pub fn unsafe_validate_cdd_expiry_nominators(targets: Vec<T::AccountId>) -> DispatchResult {
+    fn unchecked_remove_expired_cdd_nominators(targets: &Vec<T::AccountId>) -> DispatchResult {
```

Comment:
This function should make sure that the cdd has actually expired. We shouldn't be trusting a single validator unilaterally. Also, offchain tx can be submitted by anyone, I think.

Reply:
FYI - offchain txn only be submitted by the validator only. I did not add the recheck to save the cost as the validator already computed it off-chain

Reply:
I don't think that's true. I think any node can run an offchain worker and submit offchain tx.

Reply:
Yes, you are right any node can run off-chain tx but in our code, we only allowed to run by the validator as we have a check `is_validator()` that allows only validators to compute the data.

Reply:
Except that's in the tx creation phase, not propagation/import phase. therefore, anyone can artificially create the tx and then propograte. The check will not be called again.

Reply:
I think, in any case, we should double-check that each proposed nominator really does have an expired CDD check (rather than trusting the submitter, even if it can only be submitted by a validator).

Otherwise, a malicious validator could drop the CDD checks for other validators etc. and cause havoc.